### PR TITLE
Implement magnet url handling

### PR DESCRIPTION
**Pull request by felfert**
Subject says it all.

Tested with:
* ruTorrent 3.8
* transmission 2.94
* deluge 1.3.15

### Notes
While going thru the code of the adapters, I noticed some deficiencies:
Error handling is generally almost nonexistent.
The deluge implementation is quite incomplete. It should do the following.
1. Send login request
2. If status != 200 => connect error
3. If result != true => Invalid password
4. Send web.connected query
5. If status != 200 => connect error
6. If result != true => deluge-web not connected to deluge-daemon error
7. Send file or url
8. If status != 200 => send error
Currently, it performs only 1, 2, 7 and 8
For being able to show precise error messages to the user, I propose, that the parameter of the toplevel callback should be changed to a string:
If null: Everything is ok
If not null: Contains a reason or some hints. This could then be shown via notifications.

What do you think? (Would be a separate PR)

Cheers
 -Fritz

### CODE REVIEW

File: background/adapter/deluge.js [hunk 1]
```diff
@@ -30,28 +30,40 @@ torrentToWeb.adapter.deluge = function (baseUrl, username, password, autostart)
...
-                    params: [filename, window.btoa(fileReader.result), options],
+                    method: 'core.add_torrent_magnet',
+                    params: [filenameOrUrl, options],
                 };
 
                 login(() => {
                     sendRequest(requestData, callback, null);
                 });
-            });
+            } else {
```
Comment (DASPRiD) [inline background/adapter/deluge.js]: No need for an else block here, early return is preferred :) – Same applies to the other adapters.

**Comment by DASPRiD** [comment]
About your error handling: Yep, definitely makes sense. That adapter was contributed by someone, and I didn't test it myself.

File: background/adapter/deluge.js (hunk 1 above)
Comment (felfert) [reply background/adapter/deluge.js]: > No need for an else block here, early return is preferred :) – Same applies to the other adapters.

Huh? Read the code again. There is no early return possible here!
Pseudocode:

if (isMagnet) {
  handleMagnet;
} **else** {
  handleTorrent;
}

Comment (felfert) [reply background/adapter/deluge.js]: Or do you want an explicit return after handleMagnet?

**Comment by felfert** [comment]
Ok, got it. Will do it for transmission and deluge, but in rutorrent it **really** doesn't make sense.

**Comment by felfert** [comment]
> About your error handling: Yep, definitely makes sense. That adapter was contributed by someone, and I didn't test it myself.
What about my proposal to change the parameter of the success callback to a string. This is a more generic one. So would you accept a PR that changes this (for all adapters of course)?

**Comment by DASPRiD** [comment]
I'd say to hold off on that. I guess it'd make the most sense to first refactor the structure a little bit (using classes instead, using async/await, …). Then we can just throw specific errors which can be easily caught.

**Comment by felfert** [comment]
In that case, I'd love to see a new release **now** (including the Referrer and magnet PR) and start refactoring after that.

**Comment by DASPRiD** [comment]
Sounds like a plan. I'll prepare a new release this weekend. As for the refactoring: If we are going to make such a leap, it might be worthwhile going with TypeScript to get some nice typehinting in the same run.

**Comment by felfert** [comment]
BTW: This fixes #8

**Pull request merged**